mlflow/tracing/provider.py

```diff
@@ -87,34 +91,66 @@ def _setup_tracer_provider():
 
     tracer_provider = TracerProvider()
     tracer_provider.add_span_processor(processor)
+    _force_set_otel_tracer_provider(tracer_provider)
+
+
+def _force_set_otel_tracer_provider(tracer_provider):
+    """
+    Resetting internal flag used in OpenTelemetry. If we don't reset the flag,
+    set_tracer_provider() will be a no-op after the first call
+    in the application lifecycle.
+    """
+    with trace._TRACER_PROVIDER_SET_ONCE._lock:
+        trace._TRACER_PROVIDER_SET_ONCE._done = False
     trace.set_tracer_provider(tracer_provider)
 
 
 def disable():
     """
     Disable tracing by setting the global tracer provider to NoOpTracerProvider.
     """
-    if isinstance(trace.get_tracer_provider(), trace.NoOpTracerProvider):
+    if not _is_enabled():
         _logger.info("Tracing is already disabled")
         return
-
-    with trace._TRACER_PROVIDER_SET_ONCE._lock:
-        trace._TRACER_PROVIDER_SET_ONCE._done = False
-
-    trace.set_tracer_provider(trace.NoOpTracerProvider())
+    reset_tracer_setup()  # Force re-initialization of the tracer provider
+    _TRACER_PROVIDER_INITIALIZED.do_once(lambda: _setup_tracer_provider(disabled=True))
 
 
 def enable():
     """
     Enable tracing by setting the global tracer provider to the actual tracer provider.
     """
-    from mlflow.tracing.provider import _setup_tracer_provider
-
-    if not isinstance(trace.get_tracer_provider(), trace.NoOpTracerProvider):
+    if _is_enabled():
         _logger.info("Tracing is already enabled")
         return
 
-    with trace._TRACER_PROVIDER_SET_ONCE._lock:
-        trace._TRACER_PROVIDER_SET_ONCE._done = False
-
     _setup_tracer_provider()
+
+
+def reset_tracer_setup():
+    """
+    Reset the flags that indicates whether the tracer provider has been initialized.
+    This ensures that the tracer provider is re-initialized when next tracing
+    operation is performed.
+    """
+    with _TRACER_PROVIDER_INITIALIZED._lock:
+        _TRACER_PROVIDER_INITIALIZED._done = False
+    # Set NoOp tracer provider to reset the global tracer to the initial state.
+    # Do not flip _TRACE_PROVIDER_INITIALIZED flag to True so that
+    # the next tracing operation will re-initialize the provider.
+    _setup_tracer_provider(disabled=True)
+
+
+def _is_enabled() -> bool:
+    """
+    Check if tracing is enabled based on whether the global tracer
+    is instantiated or not.
+    """
+    with _TRACER_PROVIDER_INITIALIZED._lock:
+        tracer = trace.get_tracer(__name__)
+
+        # Occasionally ProxyTracer instance wraps the actual tracer
+        if isinstance(tracer, trace.ProxyTracer):
+            tracer = tracer._tracer
```

Comment:
Could you add link to where this could happen? It's hard for me to know whether this makes sense or not 🤣

Reply:
Hmm the proxy trace is returned when `OTEL_PYTHON_TRACER_PROVIDER` is not set ([code](https://github.com/open-telemetry/opentelemetry-python/blob/d73593d1137a3854ceff3d7c94180d2bdb8b097f/opentelemetry-api/src/opentelemetry/trace/__init__.py#L533-L536)), but how it is set seems to be a rabbit hole:(